digits/inference/tasks/inference.py

```diff
@@ -0,0 +1,189 @@
+        if match:
+            self.progress = float(match.group(1))/int(match.group(2))
+            return True
+
+        # path to inference data
+        match = re.match(r'Saved data to (.*)', message)
+        if match:
+            self.inference_data_filename = match.group(1).strip()
+            return True
+
+        return False
+
+    @override
+    def after_run(self):
+        super(InferenceTask, self).after_run()
+
+        # retrieve inference data
+        inputs = None
+        visualizations = []
+        outputs = {}
+        if self.inference_data_filename is not None:
+            # the HDF5 database contains:
+            # - input images, in a dataset "/inputs"
+            # - all network outputs, in a group "/outputs/"
+            # - layer activations and weights, if requested, in a group "/layers/"
+            db = h5py.File(self.inference_data_filename, 'r')
+
+            # collect inputs
+            inputs = db['inputs'][...]
+
+            # collect outputs
+            for output_name, output_data in db['outputs'].items():
+                outputs[output_name] = output_data[...]
+
+            # collect layer data, if applicable
+            if 'layers' in db.keys():
+                for layer_id, layer in db['layers'].items():
+                    visualization = {
+                        'id': int(layer_id),
+                        'name': layer.attrs['name'],
+                        'vis_type': layer.attrs['vis_type'],
+                        'data_stats': {
+                            'shape': layer.attrs['shape'],
+                            'mean': layer.attrs['mean'],
+                            'stddev': layer.attrs['stddev'],
+                            'histogram': [
+                                layer.attrs['histogram_y'].tolist(),
+                                layer.attrs['histogram_x'].tolist(),
+                                layer.attrs['histogram_ticks'].tolist(),
+                                ]
+                        }
+                    }
+                    if 'param_count' in layer.attrs:
+                        visualization['param_count'] = layer.attrs['param_count']
+                    if 'layer_type' in layer.attrs:
+                        visualization['layer_type'] = layer.attrs['layer_type']
+                    vis = layer[...]
+                    if vis.shape[0] > 0:
+                        visualization['image_html'] = embed_image_html(vis)
+                    visualizations.append(visualization)
+                # sort by layer ID (as HDF5 ASCII sorts)
+                visualizations = sorted(visualizations,key=lambda x:x['id'])
+            db.close()
+        self.inference_log.close()
+
+        # save inference to data for further use
+        self.inference_inputs = inputs
+        self.inference_outputs = outputs
+        self.inference_layers = visualizations
+
+    @override
+    def offer_resources(self, resources):
```

Comment:
Can we reserve a GPU for inference, since we are in fact using one? That seems logical, and would have saved me from some confusing out-of-memory errors. I think you can tell pycaffe which one you want to use.

It'd be cool if we could fall back to CPU-only if there are no GPUs available, so you don't have to wait for all the long-running training jobs to finish first. We could still use the `inference_task_pool` resource to limit the number of CPU inference tasks running at once.